File: server/backend/src/backend/api/models/movie.py
from __future__ import annotations

from datetime import date, datetime
from typing import Any, Optional

from pydantic import AwareDatetime, Field, computed_field

from backend.api.models.base import ResponseBase, RequestBase
from backend.api.models.movie_extra import MovieExtraAPICreate, MovieExtraAPIRead
from backend.utils.helpers import generate_uuid


# ---------- Strict input (create/update) ----------
class _MovieAPIBaseIn(RequestBase):
    """Fields for requests: validate here if needed."""

    # Fields in the media_items table
    title: str
    extended_title: Optional[str] = None
    description: Optional[str] = None
    downloaded_date: Optional[AwareDatetime] = None
    duration: float = 0
    background_image_path: Optional[str] = None
    thumbnail_landscape_path: Optional[str] = None
    thumbnail_portrait_path: Optional[str] = None
    thumbnail_square_path: Optional[str] = None
    sharing_url: Optional[str] = None
    author_name: Optional[str] = None
    author_slug: Optional[str] = None
    logo_image_path: Optional[str] = None
    mature_rating: Optional[str] = None
    has_video: bool = False
    is_downloadable: Optional[bool] = True
    status: Optional[str] = None
    published_at: Optional[AwareDatetime] = None
    background: Optional[str] = None
    byline: Optional[str] = None
    language: Optional[str] = None
    origin_country: Optional[str] = None
    images: dict[str, Any] = Field(default_factory=dict)
    available_for: list[str] = Field(default_factory=list)
    cast_and_crew: list[dict[str, Any]] = Field(default_factory=list)
    directed_by: list[str] = Field(default_factory=list)
    genres: list[Any] = Field(default_factory=list)
    hosts: list[dict[str, Any]] = Field(default_factory=list)
    more_like_this: list[dict[str, Any]] = Field(default_factory=list)
    production_companies: list[Any] = Field(default_factory=list)
    shop_items: list[Any] = Field(default_factory=list)
    starring: list[str] = Field(default_factory=list)
    written_by: list[str] = Field(default_factory=list)


class MovieAPICreate(_MovieAPIBaseIn):
    """Request body for creating a movie."""

    slug: str
    movie_extras: list[MovieExtraAPICreate] = Field(default_factory=list)
    # The extra rows do not have database IDs until this request is persisted,
    # so create requests identify the official trailer by its stable DW slug.
    official_trailer_slug: Optional[str] = None

    @computed_field(return_type=str)
    @property
    def uuid(self) -> str:
        return generate_uuid()


class MovieAPIUpdate(_MovieAPIBaseIn):
    """Request body for updating a movie."""
    pass


# ---------- Lenient output (read) ----------
class _MovieAPIBaseOut(ResponseBase):
    """Fields for responses: no validators, no constraints."""

    id: int
    uuid: str
    slug: str
    title: str
    extended_title: Optional[str]
    description: Optional[str]
    downloaded_date: Optional[datetime]
    duration: float
    background_image_path: Optional[str]
    thumbnail_landscape_path: Optional[str]
    thumbnail_portrait_path: Optional[str]
    thumbnail_square_path: Optional[str]
    sharing_url: Optional[str]
    author_name: Optional[str]
    author_slug: Optional[str]
    logo_image_path: Optional[str]
    mature_rating: Optional[str]
    has_video: bool
    is_downloadable: Optional[bool]
    status: Optional[str]
    published_at: Optional[datetime]
    background: Optional[str]
    byline: Optional[str]
    language: Optional[str]
    origin_country: Optional[str]
    images: dict[str, Any]
    available_for: list[str]
    cast_and_crew: list[dict[str, Any]]
    directed_by: list[str]
    genres: list[Any]
    hosts: list[dict[str, Any]]
    more_like_this: list[dict[str, Any]]
    production_companies: list[Any]
    shop_items: list[Any]
    starring: list[str]
    written_by: list[str]
    release_date: Optional[date]
    release_date_source: Optional[str]
    release_date_source_id: Optional[str]
    release_date_lookup_status: str
    release_date_lookup_attempted_at: Optional[datetime]
    release_date_lookup_error: Optional[str]
    official_trailer_id: Optional[int]
    official_trailer: Optional[MovieExtraAPIRead]
    movie_extras: list[MovieExtraAPIRead]


class MovieAPIRead(_MovieAPIBaseOut):
    """Represents a movie summary/detail item returned by the API."""

    created_at: datetime
    updated_at: datetime

File: ui/src/types/schemas/movie.ts
import {z} from "zod";
import {MovieExtraTypeSchema} from './dailywire_catalog'
import {ApiDateTimeSchema} from './datetime'


export const MovieExtraCreateSchema = z.object({
    slug: z.string(),
    title: z.string(),
    movieExtraType: MovieExtraTypeSchema,
    description: z.string().nullable().optional(),
    downloadedDate: z.date().nullable().optional(),
    duration: z.number().default(0),
    backgroundImagePath: z.string().nullable().optional(),
    thumbnailLandscapePath: z.string().nullable().optional(),
    thumbnailPortraitPath: z.string().nullable().optional(),
    thumbnailSquarePath: z.string().nullable().optional(),
    sharingUrl: z.string().nullable().optional(),
    publishedDate: z.date().nullable().optional(),
    availableFor: z.array(z.string()).default([]),
})
export type MovieExtraCreateIn = z.input<typeof MovieExtraCreateSchema>;
export type MovieExtraCreateOut = z.output<typeof MovieExtraCreateSchema>;


export const MovieExtraReadSchema = z.looseObject({
    id: z.int(),
    movieId: z.int(),
    uuid: z.string(),
    slug: z.string(),
    title: z.string(),
    movieExtraType: MovieExtraTypeSchema,
    description: z.string().nullable(),
    downloadedDate: ApiDateTimeSchema.nullable(),
    duration: z.number(),
    backgroundImagePath: z.string().nullable(),
    thumbnailLandscapePath: z.string().nullable(),
    thumbnailPortraitPath: z.string().nullable(),
    thumbnailSquarePath: z.string().nullable(),
    sharingUrl: z.string().nullable(),
    publishedDate: ApiDateTimeSchema.nullable(),
    availableFor: z.array(z.string()),
    createdAt: ApiDateTimeSchema,
    updatedAt: ApiDateTimeSchema,
})
export type MovieExtraRead = z.infer<typeof MovieExtraReadSchema>;

const MovieBaseSchema = z.object({
    title: z.string(),
    extendedTitle: z.string().nullable().optional(),
    description: z.string().nullable().optional(),
    downloadedDate: z.date().nullable().optional(),
    duration: z.number().default(0),
    backgroundImagePath: z.string().nullable().optional(),
    thumbnailLandscapePath: z.string().nullable().optional(),
    thumbnailPortraitPath: z.string().nullable().optional(),
    thumbnailSquarePath: z.string().nullable().optional(),
    sharingUrl: z.string().nullable().optional(),
    authorName: z.string().nullable().optional(),
    authorSlug: z.string().nullable().optional(),
    logoImagePath: z.string().nullable().optional(),
    matureRating: z.string().nullable().optional(),
    hasVideo: z.boolean().default(false),
    isDownloadable: z.boolean().nullable().optional(),
    status: z.string().nullable().optional(),
    publishedAt: z.date().nullable().optional(),
    background: z.string().nullable().optional(),
    byline: z.string().nullable().optional(),
    language: z.string().nullable().optional(),
    originCountry: z.string().nullable().optional(),
    images: z.record(z.string(), z.unknown()).default({}),
    availableFor: z.array(z.string()).default([]),
    castAndCrew: z.array(z.record(z.string(), z.unknown())).default([]),
    directedBy: z.array(z.string()).default([]),
    genres: z.array(z.unknown()).default([]),
    hosts: z.array(z.record(z.string(), z.unknown())).default([]),
    moreLikeThis: z.array(z.record(z.string(), z.unknown())).default([]),
    productionCompanies: z.array(z.unknown()).default([]),
    shopItems: z.array(z.unknown()).default([]),
    starring: z.array(z.string()).default([]),
    writtenBy: z.array(z.string()).default([]),
})

export const MovieCreateSchema = MovieBaseSchema.extend({
    slug: z.string(),
    movieExtras: z.array(MovieExtraCreateSchema).default([]),
    officialTrailerSlug: z.string().nullable().optional(),
})
export type MovieCreateIn = z.input<typeof MovieCreateSchema>;
export type MovieCreateOut = z.output<typeof MovieCreateSchema>;

export const MovieUpdateSchema = MovieBaseSchema.extend({})
export type MovieUpdateIn = z.input<typeof MovieUpdateSchema>;
export type MovieUpdateOut = z.output<typeof MovieUpdateSchema>;

export const MovieReadSchema = z.looseObject({
    id: z.int(),
    uuid: z.string(),
    slug: z.string(),
    title: z.string(),
    extendedTitle: z.string().nullable().optional(),
    description: z.string().nullable().optional(),
    downloadedDate: ApiDateTimeSchema.nullable().optional(),
    duration: z.number(),
    backgroundImagePath: z.string().nullable(),
    thumbnailLandscapePath: z.string().nullable(),
    thumbnailPortraitPath: z.string().nullable(),
    thumbnailSquarePath: z.string().nullable(),
    sharingUrl: z.string().nullable(),
    authorName: z.string().nullable(),
    authorSlug: z.string().nullable(),
    logoImagePath: z.string().nullable(),
    matureRating: z.string().nullable(),
    hasVideo: z.boolean(),
    isDownloadable: z.boolean().nullable(),
    status: z.string().nullable(),
    publishedAt: ApiDateTimeSchema.nullable(),
    background: z.string().nullable(),
    byline: z.string().nullable(),
    language: z.string().nullable(),
    originCountry: z.string().nullable(),
    images: z.record(z.string(), z.unknown()),
    availableFor: z.array(z.string()),
    castAndCrew: z.array(z.record(z.string(), z.unknown())),
    directedBy: z.array(z.string()),
    genres: z.array(z.unknown()),
    hosts: z.array(z.record(z.string(), z.unknown())),
    moreLikeThis: z.array(z.record(z.string(), z.unknown())),
    productionCompanies: z.array(z.unknown()),
    shopItems: z.array(z.unknown()),
    starring: z.array(z.string()),
    writtenBy: z.array(z.string()),
    // Calendar dates stay calendar dates. Do not invent midnight or a timezone.
    releaseDate: z.iso.date().nullable(),
    releaseDateSource: z.string().nullable(),
    releaseDateSourceId: z.string().nullable(),
    releaseDateLookupStatus: z.string(),
    releaseDateLookupAttemptedAt: ApiDateTimeSchema.nullable(),
    releaseDateLookupError: z.string().nullable(),
    officialTrailerId: z.int().nullable(),
    officialTrailer: MovieExtraReadSchema.nullable(),
    movieExtras: z.array(MovieExtraReadSchema),
    createdAt: ApiDateTimeSchema,
    updatedAt: ApiDateTimeSchema,
})
export type MovieRead = z.infer<typeof MovieReadSchema>;
